framework/src/headers/icmp.rs

```diff
@@ -0,0 +1,149 @@
+use headers::ip::IpHeader;
+use std::default::Default;
+use super::EndOffset;
+use std::marker::PhantomData;
+use num::FromPrimitive;
+use std::fmt;
+
+/*
+   ICMPv6 messages are contained in Ipv6 packets. The IPV6packet contains an IPv6 header followed by the
+   payload which contains the ICMPv6 message.
+
+   From (https://tools.ietf.org/html/rfc4443)
+   The ICMPv6 messages have the following general format:
+
+   0                   1                   2                   3
+   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |     Type      |     Code      |          Checksum             |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                                                               |
+   +                         Message Body                          +
+   |                                                               |
+
+   The type field indicates the type of the message.  Its value
+   determines the format of the remaining data.
+
+   The code field depends on the message type.  It is used to create an
+   additional level of message granularity.
+
+   The checksum field is used to detect data corruption in the ICMPv6
+   message and parts of the IPv6 header.
+*/
+
+#[derive(FromPrimitive, Debug, PartialEq)]
+#[repr(u16)]
+pub enum IcmpMessageType {
+    RouterAdvertisement = 134,
+    NeighborSolicitation = 135,
+    NeighborAdvertisement = 136,
+}
+
+impl fmt::Display for IcmpMessageType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            IcmpMessageType::RouterAdvertisement => write!(f, "Router Advertisement"),
+            IcmpMessageType::NeighborSolicitation => write!(f, "Neighbor Solicitation"),
+            IcmpMessageType::NeighborAdvertisement => write!(f, "Neighbor Advertisement"),
+        }
+    }
+
+}
+
+#[derive(Debug)]
+#[repr(C, packed)]
+pub struct IcmpV6Header<T> {
+    msg_type: u8,
+    code: u8,
+    checksum: u16,
+    _parent: PhantomData<T>,
+}
+
+impl<T> Default for IcmpV6Header<T>
+    where
+        T: IpHeader,
+{
+    fn default() -> IcmpV6Header<T> {
+        IcmpV6Header {
+            msg_type: 0,
+            code: 0,
+            checksum: 0,
+            _parent: PhantomData
+        }
+    }
+}
+
+impl<T> fmt::Display for IcmpV6Header<T>
+    where
+        T: IpHeader,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "msg_type: {} code: {} checksum: {}",
+            self.msg_type().unwrap(),
+            self.code(),
+            self.checksum()
+        )
+    }
+}
+
+
+impl<T> EndOffset for IcmpV6Header<T>
+    where
+        T: IpHeader,
+{
+    type PreviousHeader = T;
+
+    #[inline]
+    fn offset(&self) -> usize {
+        54
```

Comment:
@scrollins, probably want to comment on how we get to this offset here; Probably worth comments on each of these functions too.

Reply:
i misunderstood and thought the offset was the starting byte in the packet for the icmpv6 header. The offset should be 4. I put comments in the offset and size methods to show how those are calculated. at the top of the icmp header rust src file , i have what the rfc is for icmpv6 messages